### PR TITLE
implement better retries for tip-distributor workflows

### PR DESCRIPTION
**Pull request by segfaultdoc**
- Refreshes blockhash in between retry iterations.
- Increases retry duration

### CODE REVIEW

File: tip-distributor/src/claim_mev_workflow.rs [hunk 1]
```diff
@@ -131,21 +133,19 @@ pub fn claim_mev_tips(
             }
         }
 
-        let blockhash = rpc_client.get_latest_blockhash().await.expect("fetch latest blockhash");
         let transactions = instructions.into_iter().map(|ix|{
-            Transaction::new_signed_with_payer(
+            Transaction::new_with_payer(
```
Comment (buffalu) [inline tip-distributor/src/claim_mev_workflow.rs:137]: why not signed?

Comment (buffalu) [reply tip-distributor/src/claim_mev_workflow.rs:137]: oh nvmd sign with blockhash below

File: tip-distributor/src/lib.rs [hunk 2]
```diff
@@ -455,59 +455,59 @@ pub fn derive_tip_distribution_account_address(
...
-                Err(e) => {
-                    error!(
-                        "error confirming signature: {:?}, signature: {:?}",
-                        signature, e
-                    );
-                }
-            }
-        }
+        let futs = transactions.iter().map(|tx| async {
+            let result = rpc_client.send_and_confirm_transaction(tx).await;
```
[the remark this answers is not in the document]
Comment (buffalu) [reply tip-distributor/src/lib.rs:489]: this does:
send 1
confirm 1
send 2
confirm 2

imo should:
send 1
send 2
...
confirm 1
confirm 2
...

[the remark this answers is not in the document]
Comment (buffalu) [reply tip-distributor/src/lib.rs:489]: oh i gues this is a bunch of futs, so maybe not too bad

[the remark this answers is not in the document]
Comment (buffalu) [reply tip-distributor/src/lib.rs:489]: how long does confirm wait by default?

[the remark this answers is not in the document]
Comment (segfaultdoc) [reply tip-distributor/src/lib.rs:489]: loops until it confirms, fails or blockhash expires

File: tip-distributor/src/lib.rs [hunk 3]
```diff
@@ -455,59 +455,59 @@ pub fn derive_tip_distribution_account_address(
...
+            .map(|(sig, result)| {
+                let e = result.err().unwrap();
+                warn!("error sending transaction: [error={e}, signature={sig}]");
+                (sig, e)
+            })
+            .collect::<HashMap<_, _>>();
+
+        transactions = errors
+            .iter()
+            .map(|(sig, _)| signatures_to_transactions.remove(sig).unwrap())
```
Comment (buffalu) [inline tip-distributor/src/lib.rs:506]: you wanna remove the succesful ones, not the errored ones

Comment (segfaultdoc) [reply tip-distributor/src/lib.rs:506]: we wanna retry the failed ones

Comment (segfaultdoc) [reply tip-distributor/src/lib.rs:506]: can move the hashmap outside the loop then we'd rm only the succeeded ones and clone the errored ones for retry

File: tip-distributor/src/lib.rs [hunk 4]
```diff
@@ -455,59 +455,70 @@ pub fn derive_tip_distribution_account_address(
...
-        .map(|tx| (tx.signatures[0], tx.clone()))
-        .collect();
+    mut transactions: Vec<Transaction>,
+    max_retry_duration: Duration,
+) -> HashMap<Signature, Error> {
+    let mut errors = HashMap::default();
+    let mut blockhash = rpc_client
+        .get_latest_blockhash()
+        .await
+        .expect("fetch latest blockhash");
```
Comment (buffalu) [inline tip-distributor/src/lib.rs:468]: you should be able to sign here + on blockhash refresh in main loop. doing a bunch of resigns messy

File: tip-distributor/src/lib.rs [hunk 5]
```diff
@@ -455,59 +455,70 @@ pub fn derive_tip_distribution_account_address(
...
-                }
-                Ok(false) => {
-                    info!("didn't confirmed signature: {:?}", signature);
-                }
-                Err(e) => {
-                    error!(
-                        "error confirming signature: {:?}, signature: {:?}",
-                        signature, e
-                    );
+        let futs = transactions.iter().map(|tx| async {
```
Comment (buffalu) [inline tip-distributor/src/lib.rs:488]: lots of copying between this + signatures_to_transactions, should be able to just use signatures_to_transactions and avoid the juggling of two variables

File: tip-distributor/src/lib.rs (hunk 3 above)
Comment (segfaultdoc) [reply tip-distributor/src/lib.rs]: the commit that your commenting on wasn't using the `signatures_to_transactions` as a retry map and was getting initialized on each iteration, new code is different

**Pull request merged**